File: 400. Nth Digit/Java/Solution.java
public class Solution {
    public int findNthDigit(int n) {
        int digits = 1;
        int start = 1;
        long length = 9;
        while (n > digits * length) {
            n -= digits * length;
            digits++;
            start *= 10;
            length *= 10;
        }
        n--;
        int number = start + n / digits;
        return Character.getNumericValue(String.valueOf(number).charAt(n % digits));
    }
}

File: 400. Nth Digit/Python/Solution.py
class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        digits = 1
        start = 1
        length = 9
        while (n > digits * length):
            n -= digits * length
            digits += 1
            start *= 10
            length *= 10
        n -= 1
        number = start + n // digits
        return int(str(number)[n % digits])
